Fix  error checking and bytecode circuit

adria0 opened the pull request:
With `ok()` errors are turned into an `Option` so the compiler do not check the result and also do not warns about unchecked result.

It seems that this woke up test errors due to changes in the underlying halo2 in some moment ( see https://github.com/appliedzkp/zkevm-circuits/issues/407#issuecomment-1081969840 ), so also applied @Brechtpd workaround.

Commit:
Fix `Error`s converted into `Option`s, not checking the return error

ed255 commented:
Good catch!  And it seems there were indeed silent errors on the bytecode circuit tests!  Probably the `k` values in the tests in `zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs` need to be bumped.

CPerezz commented:
The change looks good! It's a nice catch!

Before approving, I would like that the tests that are now failing are fixed.

The `NotEnoughRowsAvaliable` err suggests that the `k` param passed to the MockProver in the `run` fn is lower than what's actually required. My suggestion is to make these values bigger so that they fit the size requirements.

CPerezz reviewed (changes requested):
The change looks good! It's a nice catch!

Before approving, I would like that the tests that are now failing are fixed.

The `NotEnoughRowsAvaliable` err suggests that the `k` param passed to the MockProver in the `run` fn is lower than what's actually required. My suggestion is to make these values bigger so that they fit the size requirements.

adria0 commented:
@ed255 @CPerezz 

Wondering if anything else wrong there, I'm running the test, with only one byte of code and one million rows `k=20` and still needs more

```
/// Tests a circuit with incomplete bytecode
717     #[test]
718     fn bytecode_incomplete() {
719         let k = 20;
720         let r = MyCircuit::r();
721         verify::<Fr>(k, vec![unroll(vec![7u8; 1], r)], false);
```

fails with

```
[zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs:618] config.assign(layouter, self.size, &self.bytecodes) = Err(
    NotEnoughRowsAvailable {
        current_k: 20,
    },
)
thread 'bytecode_circuit::bytecode_unroller::tests::bytecode_incomplete' panicked at 'called `Result::unwrap()` on an `Err` value: NotEnoughRowsAvailable { current_k: 20 }', zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs:629:64
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

CPerezz commented:
The error definition in Halo2 is quite clear:
```rust
/// `k` is too small for the given circuit.
    NotEnoughRowsAvailable {
        /// The current value of `k` being used.
        current_k: u32,
    },
```

Is it possible that you're passing more `instances` than required?

therealyingtong commented:
I think this fixes the `MockProver` error: https://github.com/appliedzkp/zkevm-circuits/issues/407#issuecomment-1078988923

CPerezz commented:
> I think this fixes the `MockProver` error: [#407 (comment)](https://github.com/appliedzkp/zkevm-circuits/issues/407#issuecomment-1078988923)

Sorry we didn't update earlier here @therealyingtong. 
In fact, this solved the error. But the circuit is still not passing due to other things (constraints not satisfied etc...).

Commit:
Convert `q_enable` from Selector to Fixed, as proposed by @brechtpd in
https://github.com/appliedzkp/zkevm-circuits/issues/407#issuecomment-1081969840

Commit:
fmt

adria0 committed:
Merge branch 'main' into fix/fill_table_err_check

therealyingtong commented:
While reviewing the `bytecode_unroller` circuit I had a few questions / comments:

- `is_code` should be boolean-constrained (like `is_final` and `padding` are);
- why do we need to witness `hash`, `hash_length` on every row, if it's only used on the row where `is_final == 1`?

adria0 commented:
> While reviewing the `bytecode_unroller` circuit I had a few questions / comments:

thanks for reviewing. :)

> * `is_code` should be boolean-constrained (like `is_final` and `padding` are);

Yes, you're right!

> * do we expect the `bytecode_invalid_index` test to fail?

As I understood well the test code, the idea was to deliver more unrolled code to the circuit that can handle (and this fails with a `low k` panic!). Maybe in a previous halo2 version used the library does not paniked and returned a Result::Err, but for now I have to catch this panic!. Or maybe I misunderstood the reason for this test?

BTW, I saw your previous comment about witnessing unqueried values, and I was checking if it works in a way do you said. Just for curiosity why do removed this comment?

therealyingtong commented:
I think I see why we're currently catching a panic for the `bytecode_incomplete` test; but why are we ignoring the `bytecode_invalid_index` test?

Brechtpd commented:
> `is_code` should be boolean-constrained (like `is_final` and `padding` are);

It's already being constrained in both the start/continue cases so should be fine I think?

> but why are we ignoring the bytecode_invalid_index test?

Oh right I forgot about that, I temporarily disabled that test because halo2 gave this internal error only when providing incorrect witness values in the test: `thread 'bytecode_circuit::bytecode_unroller::tests::bytecode_invalid_index' panicked at 'internal error: entered unreachable code', /halo2/halo2_proofs/src/dev/util.rs:48:42:`. This happened after some update was done to halo2 because while I was writing these tests it worked correctly (as in the circuit correctly returned an error for the incorrect witness).

> I had previously asked why we witnessed hash, hash_length on every row, when we're only looking it up on rows where is_final == 1.

We'll need it on each row for external circuits using the data in lookups.

CPerezz commented on zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs:
```diff
@@ -715,6 +716,7 @@ mod tests {
 
     /// Tests a circuit with incomplete bytecode
     #[test]
+    #[should_panic = "called `Result::unwrap()` on an `Err` value: NotEnoughRowsAvailable { current_k: 9 }"]
```
If it's incomplete how can it panic due to `NotEnoughRowsAvailable`??

adria0 replied:
AFAIU `incomplete` here means that not all code can be processed, but probally there's another definition that I'm missing.

adria0 replied:
@Brechtpd, I see, thanks!

Yes, I can make the circuit failing as you said by not adding rows when assigning witness if they are going to overflow the table, and works,

```
341                        if offset <= last_row_offset {
342                             // Set the data for this row
343                             self.set_row(
```

Nonetheless, I'm asking myself if creating witness that makes a circuit failing a good strategy for tests. Is not just better to fail with `NotEnoughRowsAvailable`? 

Anyway we are going to change how bytecode circuits works, so maybe we do not have to spend more time on this issue :)

Brechtpd replied:
> Nonetheless, I'm asking myself if creating witness that makes a circuit failing a good strategy for tests. Is not just better to fail with NotEnoughRowsAvailable?

The main goal of the circuits is to reject incorrect data, and so the only way to really test that is to check if the circuit indeed cannot be verified with that data. The fact that halo2 library throws an error assigning the witness is actually not useful from a security pov because that is something a malicious prover can just work around by changing a bit of code when he generates the proof.

> Anyway we are going to change how bytecode circuits works, so maybe we do not have to spend more time on this issue :)

Yeah true :), but I think the tests can remain the same because the change is just how the circuit does some of the constraints, and so the tests still need to test the same things.

adria0 replied:
> The main goal of the circuits is to reject incorrect data, and so the only way to really test that is to check if the circuit indeed cannot be verified with that data. The fact that halo2 library throws an error assigning the witness is actually not useful from a security pov because that is something a malicious prover can just work around by changing a bit of code when he generates the proof.

Super good information to keep in mind! Thanks!

> Yeah true :), but I think the tests can remain the same because the change is just how the circuit does some of the constraints, and so the tests still need to test the same things.

Fixed in https://github.com/appliedzkp/zkevm-circuits/pull/403/commits/f33f782b5960233e13518f96a90e99e696b4651f (I hope!)

CPerezz reviewed (commented):
So just changing `q_enable` from `Selector` to a `Fixed` column made it?
Could you explain why is that specifically? I'm really curious!

Brechtpd commented:
> So just changing `q_enable` from `Selector` to a `Fixed` column made it? Could you explain why is that specifically? I'm really curious!

The MockProver requires any gate with a selector enabled to have all cells referenced in that gate to be assigned. Fixed columns do not have this extra check. I also find this difference a bit strange (and impossible to achieve in certain scenario's) and made this issue: https://github.com/zcash/halo2/issues/533.

Commit:
Do not assign all bytecode witness if overflows the circuit usable rows

adria0 committed:
Merge branch 'main' into fix/fill_table_err_check

adria0 commented:
@miha-stopar I did not manually required your review, just the new automatic system assigned it to you!

miha-stopar reviewed (approved):
LGTM!

ed255 commented on zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs:
```diff
@@ -307,38 +307,38 @@ impl<F: Field> Config<F> {
+                        // Add the byte to the accumulator
+                        hash_rlc = hash_rlc * self.r + row.byte;
+
+                        if offset <= last_row_offset {
```
If we don't enter this condition we know we're generating an invalid proof.  This is good for testing but it's not good outside of testing.
We have a similar situation in the state circuit, where we want to assign invalid advices for testing purposes, but we want an error in the assign function outside of tests.  In that case there's a generic const bool called [`SANITY_CHECK`](https://github.com/appliedzkp/zkevm-circuits/blob/a9bdc9a87f9fcb8068fd4a3f254c94194a0f4e1d/zkevm-circuits/src/state_circuit/state.rs#L554) that when true, makes the assign function panic on invalid witness data.  Could we have a similar implementation here?

Expanding on that, we could have two methods: `assign_unchecked` and `assign`. where the first one doesn't do sanity checks and the later does.

ed255 replied:
My idea of `assign_unchecked` and `assign` was to do the following:
```rust
    fn _assign<const SANITY_CHECK: bool>(
        &self,
        mut layouter: impl Layouter<F>,
        size: usize,
        witness: &[UnrolledBytecode<F>],
    ) -> Result<(), Error> {
        // Full Implementation with conditional sanity check
    }
    pub(crate) fn assign(
        &self,
        mut layouter: impl Layouter<F>,
        size: usize,
        witness: &[UnrolledBytecode<F>],
    ) -> Result<(), Error> {
        self._assign::<true>(layouter, size, witness)
    }
    fn assign_unchecked(
        &self,
        mut layouter: impl Layouter<F>,
        size: usize,
        witness: &[UnrolledBytecode<F>],
    ) -> Result<(), Error> {
        self._assign::<false>(layouter, size, witness)
    }
```

But I like your idea of using the `#[cfg(test)]`!  Nevertheless, using features doesn't allow us to test both options in unit tests, so we wouldn't be able to test the sanity checks in our unit test functions.  But I think that's OK!

adria0 replied:
I changed my mind a little. It really creates me some panic to have different code in production and test, and since we can activate features on tests, a better option could be:

```rust
let set_row = if cfg!(feature = "sanity_check") {
   offset <= last_row_offset
} else {
   true
};

if set_row {
   self.set_row(...
```

then in tests we could add 

```rust
/// Tests a circuit with incomplete bytecode (sanity check)
fn test_bytecode_incomplete() {
   let k = 9;
   let r = MyCircuit::r();
   verify::<Fr>(k, vec![unroll(vec![7u8; 2usize.pow(k) + 1], r)], false);
}

#[test]
#[cfg(feature = "sanity_check")]
fn bytecode_incomplete() {
   test_bytecode_incomplete();
}
#[test]
#[cfg(not(feature = "sanity_check"))]
#[should_panic]
fn bytecode_incomplete() {
   test_bytecode_incomplete();
}
```

then we can run the tests, activating the feature or not `cargo test --features=sanity_check`. The bad thing is that we have to run the tests twice.

ed255 replied:
> then we can run the tests, activating the feature or not `cargo test --features=sanity_check`. The bad thing is that we have to run the tests twice.

Not just this; we also would be introducing and exposing a feature to the library that is only meaningful for internal tests.

Also from your example, from my point of view, when the sanity check is enabled, an invalid witness (or incompatible circuit setup) should cause an error or panic.  So to me it should look like:
```rust
if cfg!(feature = "sanity_check") {
  if offset > last_row_offset {
    panic!("offset > last_row_offset");
  }
}
if offset <= last_row_offset {
  self.set_row(...
}

adria0 replied:
@ed255, with this code it will panic with "low k" is sanity_check is not set, and will panic with "offet > last_row_offset" if sanity_check is set.

I think that the main idea here, and (as I undestood from https://github.com/appliedzkp/zkevm-circuits/pull/403#discussion_r844088585) is how to create bad witnesses in a very targeted way (so, not using a generic fuzzer) to make the circuit to fail.

- Maybe we could decide if we introduce this kind of tests ( bad witnesses ) or not in our current state.
- Also (I think that we talked on this) is if we have to have a mechanism to decide when returning an error or when to panic ( to get the stacktrace ), because it also affects how "bad tests" are going yo be written. A would like some kind of retpanic!( Result::Err ) that panics or returns depending on compliation params.

ed255 replied:
After an offline discussion with @adria0 we agree that it's very necessary to test the circuits with invalid inputs, but the fact that the `assign` functions are designed to calculate valid intermediate witness makes it hard, so we'd like to think more about this and move on with this PR for now (specially considering that the bytecode circuit has a [pending refactor](https://github.com/appliedzkp/zkevm-specs/issues/151)).

ed255 replied:
I see that there's currently an open issue for negative tests on the circuits, maybe we can continue the discussion there? https://github.com/appliedzkp/zkevm-circuits/issues/400

CPerezz reviewed (commented):
I think we should follow @ed255 's approach mentioned here: https://github.com/appliedzkp/zkevm-circuits/pull/403#discussion_r844993664

adria0 committed:
Merge branch 'main' into fix/fill_table_err_check

ed255 reviewed (approved):
I'm approving this PR following the conclusion at https://github.com/appliedzkp/zkevm-circuits/pull/403#discussion_r846022555 and moving the discussion of negative tests to https://github.com/appliedzkp/zkevm-circuits/issues/400

@CPerezz does this seem OK to you? (as currently your review status is "requested changes")

adria0 committed:
Merge branch 'main' into fix/fill_table_err_check

adria0 committed:
Merge branch 'main' into fix/fill_table_err_check

Commit:
fix merge

CPerezz reviewed (approved):
LGTM.

Sorry for blocking on the merge of this.

Let's try to not forget the work left or tech-debt generated here by tracking this in https://github.com/appliedzkp/zkevm-circuits/issues/400
